### PR TITLE
EVM statedb

### PR DESCRIPTION
**Pull request by JcBernack**
- refactor libevm interface to simplify Java interop
  - this will greatly reduce development overhead when adding new methods to the library
- add StateDB class that manages an underlying instance of the go-ethereum StateDB

### CODE REVIEW

File: libevm/lib/service_state.go [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+		if end > length {
+			end = length
+		}
+		chunk := statedb.GetState(params.Address, getChunkKey(params.Key, i+1))
+		copy(data[i:end], skipLeadingZeroes(chunk))
+	}
+	return nil, data
+}
+
+// StateSetStorage writes values of arbitrary length to the storage trie of given account
```
Comment (JcBernack) [inline libevm/lib/service_state.go:220]: The algorithm for storing values larger than 32 bytes works but is not particularly pretty.

File: evm/src/main/java/com/horizen/evm/interop/EvmLog.java [hunk 2]
```diff
@@ -0,0 +1,19 @@
+package com.horizen.evm.interop;
+
+import com.horizen.evm.utils.Address;
+import com.horizen.evm.utils.Hash;
+import com.horizen.evm.utils.HexBytes;
+
+import java.math.BigInteger;
+
+public class EvmLog {
```
Comment (JcBernack) [inline evm/src/main/java/com/horizen/evm/interop/EvmLog.java:9]: Currently, this contains everything the log in Ethereum has to offer. Depending on how we create and store reciepts in the core we might be able to reduce this.

File: evm/src/main/java/com/horizen/evm/LibEvm.java [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+    public static EvmResult evmExecute(int handle, byte[] from, byte[] to, BigInteger value, byte[] input)
+        throws Exception {
+        var cfg = new EvmParams.EvmConfig();
+        cfg.origin = Address.FromBytes(from);
+        cfg.value = value;
+        cfg.gasLimit = 100000;
+        return invoke("EvmExecute", new EvmParams(handle, cfg, to, input), EvmResult.class);
+    }
+
+    public static EvmApplyResult evmApply(int handle, byte[] from, byte[] to, BigInteger value, byte[] input)
```
Comment (JcBernack) [inline evm/src/main/java/com/horizen/evm/LibEvm.java:145]: This is an alternative method to call the EVM which uses more of the code-path in GETH when applying a block. It has more validation and also handles gas fees (Intrinsic Gas + moving gas from the caller balance to the coinbase account). Which one we're going to use, or if we need something a bit different depends on how exactly we want to handle gas fees.

File: evm/src/main/java/com/horizen/evm/LibEvm.java [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+import com.horizen.evm.interop.*;
+import com.horizen.evm.utils.Address;
+import com.horizen.evm.utils.Hash;
+import com.sun.jna.Library;
+import com.sun.jna.Native;
+import com.sun.jna.Pointer;
+
+import java.math.BigInteger;
+
+final class LibEvm {
```
Comment (rockNicktive) [inline evm/src/main/java/com/horizen/evm/LibEvm.java:13]: For gas estimation, is the intention to "throw the state chages away" on the java side and simply use evmExecute?

**Review by rockNicktive — APPROVED** [review]
I've read through the code changes. I'll test it on my machine later, but so far everything is looking great and I don't expect any issues!

File: evm/src/main/java/com/horizen/evm/LibEvm.java (hunk 4 above)
Comment (JcBernack) [reply evm/src/main/java/com/horizen/evm/LibEvm.java:13]: For gas estimation I expect to re-use parts or even all of the geth implementation. So I think it makes more sense to do that on the go-library side.

File: evm/src/main/java/com/horizen/evm/interop/EvmLog.java (hunk 2 above)
Comment (JcBernack) [reply evm/src/main/java/com/horizen/evm/interop/EvmLog.java:6]: Removed all derived fields.

**Review by i-Alex — CHANGES_REQUESTED** [review]
Looks good.
The only point about the physical storage usage.

**Review by i-Alex — APPROVED** [review]

**Pull request merged**